sycl/source/detail/pi.cpp

```diff
@@ -309,14 +311,321 @@ bool trace(TraceLevel Level) {
+        // Next, match the device_type entry
+        if (FilterDevType == info::device_type::all) {
+          // Last, match the device_num entry
+          if (!Filter.HasDeviceNum || DeviceNum == Filter.DeviceNum) {
+            PiDevices[InsertIDx++] = Device;
+            break;
+          }
+        } else if (FilterDevType == DeviceType) {
+          if (!Filter.HasDeviceNum || DeviceNum == Filter.DeviceNum) {
+            PiDevices[InsertIDx++] = Device;
+            break;
+          }
+        }
+      }
+    }
+    DeviceNum++;
+  }
+  PiDevices.resize(InsertIDx);
+}
+
+// Fill up the platform cache and device cache for the given Plugin.
+// Thi<s is a one time process when a plugin is constructed.
+void fillPlatformAndDeviceCache(plugin &Plugin) {
+  pi_uint32 NumPlatforms = 0;
+  // Move to the next plugin if the plugin fails to initialize.
+  // This way platforms from other plugins get a chance to be discovered.
+  if (Plugin.call_nocheck<PiApiKind::piPlatformsGet>(
+          0, nullptr, &NumPlatforms) != PI_SUCCESS)
+    return;
+
+  if (NumPlatforms) {
+    vector_class<RT::PiPlatform> PiPlatforms(NumPlatforms);
+    if (Plugin.call_nocheck<PiApiKind::piPlatformsGet>(
+            NumPlatforms, PiPlatforms.data(), nullptr) != PI_SUCCESS)
+      return;
+
+    std::vector<PlatformImplPtr> &PlatformCache =
+        GlobalHandler::instance().getPlatformCache();
+    std::vector<DeviceImplPtr> &DeviceCache =
+        GlobalHandler::instance().getDeviceCache();
+
+    int DeviceNum = 0;
+    for (const auto &PiPlatform : PiPlatforms) {
+      PlatformImplPtr PlatformImpl =
+          std::make_shared<platform_impl>(PiPlatform, Plugin);
+      platform Platform = detail::createSyclObjFromImpl<platform>(PlatformImpl);
+
+      if (IsBannedPlatform(Platform))
+        continue;
+
+      // get devices
+      info::device_type DeviceType = info::device_type::all;
+      pi_uint32 NumDevices = 0;
+      Plugin.call<PiApiKind::piDevicesGet>(
+          PiPlatform, pi::cast<RT::PiDeviceType>(DeviceType), 0,
+          pi::cast<RT::PiDevice *>(nullptr), &NumDevices);
+
+      if (NumDevices == 0)
+        continue;
+
+      vector_class<RT::PiDevice> PiDevices(NumDevices);
+      Plugin.call<PiApiKind::piDevicesGet>(
+          PiPlatform, pi::cast<RT::PiDeviceType>(DeviceType), NumDevices,
+          PiDevices.data(), nullptr);
+
+      // Filter out devices that are not present in the allowlist
+      if (SYCLConfig<SYCL_DEVICE_ALLOWLIST>::get())
+        filterAllowList(PiDevices, PiPlatform, Plugin);
+
+      int UnfilteredDeviceCount = PiDevices.size();
+      // Filter out devices that are not compatible with SYCL_DEVICE_FILTER
+      filterDeviceFilter(PiDevices, PiPlatform, Plugin, DeviceNum);
+
+      if (PiDevices.size() != 0) {
+        const std::lock_guard<std::mutex> Guard(
+            GlobalHandler::instance().getPlatformMapMutex());
+
+        PlatformCache.emplace_back(PlatformImpl);
```

Comment:
```suggestion
    {
        const std::lock_guard<std::mutex> Guard(
            GlobalHandler::instance().getPlatformMapMutex());

        PlatformCache.emplace_back(PlatformImpl);
      }
```
Suggest reducing lock scope.